Feature/extract refactor

mitzimorris opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Refactor CmdStanMCMC object to encapsulate metadata more cleanly and to improve methods `draws`, `stan_variable`, and `stan_variables`.

Introduce class `InferenceMetadata` with the goal of improving the amount of information available in Arviz for a data imported from CmdStanPy.

Metadata is everything other than the sampler outputs:  draws, stepsize, and metric, including:

- all information contained in the `runset` attribute
- information parsed out of the Stan CSV file header, specifically all keyword = value pairs (excluding `file` which is ambiguous and captured by the runset).
- Stan CSV output header rows

For sampler outputs, we know that the header row names consist of

-  sampler variables, which by convention end in `__`, e.g.,  `lp__`.  All sampler variables are scalar.
-  Stan program variables.  Stan program variables can be either scalar or multidimensional.  For the latter, the variable name refers to the container, and the CSV column labels refer to the elements.  The CSV column labels  are of the form "name.x", "name.x.y", etc where each x and y are the index for dimensions 1 .. N, e.g. `z_forecast.1.2`.  By convention, these names are translated into proper array index notation, e.g. `z_forcast[1,2]`.  

Given this, the `InferenceMetadata` has 4 attributes:

- `cmdstan_config`
- `sampler_vars_cols`
- `stan_vars_cols`
- `stan_vars_dims`

The latter 2 items support the `stan_variable` and `stan_variables` methods.

Another item in this refactor:  added argument `concat_chains` to the `draws` method; this returns a 2D numpy.ndarray instead of a 3D numpy.ndarray, flattening the chains.  I.e., given 4 chains with 1000 draws each containing 8 outputs, the 3D array has shape (1000, 4, 8) and the 2D array has shape (4000, 8).  Underlyingly, the array elements are the same.




#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):   Columbia University


By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

mitzimorris commented:
hi @ahartikainen - still working on improving the sphinx docs for this package, but would very much appreciate your input on this refactor.

mitzimorris commented:
OK, this refactor is ready for review.

I rewrote the sections of the Sphinx documentation on sampling.  Since the docs were first written, we changed the interface so that models are automatically compiled during intantiation; removed stray calls to model `compile` method.

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+- ``metric`` - Per-chain metric values, either a vector or matrix. 
+- ``step_size`` - Per-chain stepsize.
+  
+- ``draws`` - A numpy.ndarray which contains all across all chains arranged as (raws, chains, columns).
```
1. These are the constrained variables?
2. The column names are in `column_names`?
3. I can figure out which columns to find a variable in with `stan_vars_cols`?

Also raws -> draws

mitzimorris replied:
yes, the constrained variables - that's what the compiled model's `write_array` method does.

mitzimorris replied:
answers to 2 and 3 also yes and yes - column indexing is 0-based - that's Python

bbbales2 reviewed (commented):
I had a dig through this -- basically the behavior of the stuff documented in `sample.rst` is what needs to be checked?

ahartikainen commented on cmdstanpy/stanfit.py:
```diff
@@ -171,42 +174,34 @@ def chain_ids(self) -> List[int]:
     def _check_retcodes(self) -> bool:
-        """True when all chains have retcode 0."""
+        """Returns ``True`` when all chains have retcode 0."""
         for i in range(self._chains):
             if self._retcodes[i] != 0:
                 return False
         return True
```
```suggestion
        return all(self._retcodes)
```

There is no need to change this, but I think this does the same thing

mitzimorris replied:
flip the logic - wouldn't that be "not all"?
because all retcodes need to be 0 - the long way is easier to read and reason about.

ahartikainen commented on cmdstanpy/utils.py:
```diff
@@ -639,47 +637,72 @@ def scan_column_names(fd: TextIO, config_dict: Dict, lineno: int) -> int:
 def scan_metric(fd: TextIO, config_dict: Dict, lineno: int) -> int:
     """
-    Scan stepsize, metric from  stan_csv file comment lines,
+     Scan stepsize, metric from  stan_csv file comment lines,
```
```suggestion
    Scan stepsize, metric from  stan_csv file comment lines,
```

mitzimorris replied:
done!

ahartikainen commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+- ``metric`` - Per-chain metric values, either a vector or matrix. 
+- ``step_size`` - Per-chain stepsize.
+  
+- ``draws`` - A numpy.ndarray which contains all across all chains arranged as (raws, chains, columns).
```
```suggestion
- ``draws`` - A numpy.ndarray which contains all draws across all chains arranged as (draws, chains, columns).
```

mitzimorris replied:
fixed!

ahartikainen reviewed (commented):
Looks good.

I think this works even when algorithm=hmc? It adds a bit of different diagnostics than nuts, but the code will still work (?)

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+
+Methods for saving the CmdStan output files:
+
+- ``save_csvfiles(dir_name)`` - Move output Stan CSV files to specified directory. 
```
Should there be an underscore here in the csvfiles name? See:

```
>>> fit.save_csvfiles
<bound method CmdStanMCMC.save_csvfiles of CmdStanMCMC: model=test chains=4['method=sample', 'algorithm=hmc', 'adapt', 'engaged=1']
 csv_files:
        /tmp/tmpwxcqwf_l/test-202102090856-1-cssacms9.csv
        /tmp/tmpwxcqwf_l/test-202102090856-2-w1hcupqr.csv
        /tmp/tmpwxcqwf_l/test-202102090856-3-3uggqmdv.csv
        /tmp/tmpwxcqwf_l/test-202102090856-4-gc51buo8.csv
 output_files:
        /tmp/tmpwxcqwf_l/test-202102090856-1-cssacms9-stdout.txt
        /tmp/tmpwxcqwf_l/test-202102090856-2-w1hcupqr-stdout.txt
        /tmp/tmpwxcqwf_l/test-202102090856-3-3uggqmdv-stdout.txt
        /tmp/tmpwxcqwf_l/test-202102090856-4-gc51buo8-stdout.txt>
```

mitzimorris replied:
probably - but that was what was decided on with a lot of back and forth, so reluctant to change this now.

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -32,7 +33,7 @@ NUTS-HMC sampler configuration
 - ``save_warmup``: When ``True``, sampler saves warmup draws as part of output csv file.
 
-- ``thin``: Period between saved samples.
+- ``thin``: Period between saved samples (draws).  Default is 1, i.e., save all iterations.
```
Should this be readable at the output? `fit.thin` for instance

mitzimorris replied:
it's available in that it's in the dict returned by propery `cmdstan_config`.  trying to avoid interface code clutter.  OTOH, the more that's in the docs, the better, so will add more to docs.

mitzimorris replied:
agreed - added this as a property.

bbbales2 commented on docsrc/hello_world.rst:
```diff
@@ -82,44 +82,65 @@ which contains all messages written to the console.
+The CmdStanMCMC class has properties and methods to access the sampler
+outputs and all metadata. Accessor properties include:
+
+ * draws - all draws from all chains, stored as a 3D numpy.ndarray.
```
Summary, stan_variable, and stan_variables should go here. I think `summary` and `stan_variable` would be the two things I'd use the most with cmdstanpy.

bbbales2 commented on docsrc/api.rst:
```diff
@@ -11,11 +11,22 @@ Classes
+    runs the HMC-NUTS sampler to produce a set of draws from the posterior distribution.
 
+optimize
+    produce a penalized maximum likelihood estimate (point estimate) of the model parameters.
```
`summary` and `stanvariable` don't work with optimize. I guess the commandline summary doesn't work, so maybe that's a miss, but I think `stanvariable` should. This can be followup stuff.

bbbales2 replied:
And I wasn't able to get variational running, but I think that interface should mimic the MCMC as much as possible.

mitzimorris replied:
yes, followup PR will generalize this to all methods.
but wanted to proceed stepwise as much as possible.

mitzimorris replied:
next PR, for sure.

bbbales2 replied:
Sounds good

bbbales2 commented on docsrc/api.rst:
```diff
@@ -11,11 +11,22 @@ Classes
+optimize
+    produce a penalized maximum likelihood estimate (point estimate) of the model parameters.
 
+variational
```
I had a problem running variational inference. I'll open a separate issue for that.

bbbales2 reviewed (changes requested):
Comments! I didn't review the code. I just read the docs and tried to break stuff.

mitzimorris commented:
> I think this works even when algorithm=hmc? It adds a bit of different diagnostics than nuts, but the code will still work (?)

yes, handling of sampler parameters is general enough to handle regular HMC, although right now, CmdStanPy doesn't let you do that - for that we'd probably need to add logic to sampler args config and also csv output parsing - haven't investigated.

mitzimorris commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+
+- ``column_names`` - Column labels for one draw from the sampler.
+- ``sampler_vars_cols`` - Maps the sampler parameter names to output column indices.
+- ``stan_vars_cols`` - Maps the Stan progran variable names to output column indices.
```
right - this is confusing.  I will try to make this more clear in the writeup.

bbbales2 replied:
If `fit.sampler_vars_cols` just returns a map of length 1 tuples the behavior will be consistent between the two and no need to explain the difference.

mitzimorris replied:
sure, that's general and woud allow for sampler vars to have structure.

mitzimorris commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+- ``column_names`` - Column labels for one draw from the sampler.
+- ``sampler_vars_cols`` - Maps the sampler parameter names to output column indices.
+- ``stan_vars_cols`` - Maps the Stan progran variable names to output column indices.
+- ``stan_vars_dims`` - Maps the Stan progran variable names to dimensions; scalar variables have zero dimensions.
```
good call.  changed to descriptions in terms of dictionary of (key, value) pairs.

mitzimorris commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+Contents of the the sample: 
 
+- ``metric`` - Per-chain metric values, either a vector or matrix. 
+- ``step_size`` - Per-chain stepsize.
```
good catch - thanks!

mitzimorris commented on docsrc/sample.rst:
```diff
@@ -70,28 +71,52 @@ The :ref:`class_cmdstanmodel` class method  ``sample`` returns a ``CmdStanMCMC``
+  
+- ``draws`` - A numpy.ndarray which contains all across all chains arranged as (raws, chains, columns).
 - ``stan_variable(var_name)`` - Returns a numpy.ndarray which contains the set of draws in the sample for the named Stan program variable.
 - ``stan_variables()`` - Return dictionary of all Stan program variables.
```
not yet.  discussed with Ari - we need to document that numpy.ndarray has operation "astype(int)" which will do the right thing.   and the thing is, draws array is all same type, so default's gotta be float.

ahartikainen replied:
integer as float is safer than float as integer (this should be float until we get dtype from csv info or somewhere else)

bbbales2 replied:
Makes sense

bbbales2 commented:
Ping me when it's ready to check again. @ahartikainen did you check the code and the tests? Or just the code? I am asking to know what I should check in my second review (if you've already checked the code, I trust your review more than mine, but I can also pedantically check the tests and such if that hasn't been done).

mitzimorris commented:
made changes requested, took another pass at improving the docs, and added a few more unit tests.
ready for re-review.

mitzimorris commented:
for the record, trying to write the docs led to a whole bunch of little fixes.  also uncovered problems introduced by late-night decision to add parameter `concat_chain` to the `draws` method.  added a few more unit tests to check my work.

bbbales2 commented on docsrc/hello_world.rst:
```diff
@@ -82,44 +82,38 @@ which contains all messages written to the console.
+By default, it returns the underlying 3D array.
+The optional boolean argument ``concat_chains``, when ``True``,
+will flatten the chains resulting in a 2D array.
+the size of the draws dimension is the number of sampling iterations.
```
'the size of the draws dimension is the number of sampling iterations.' seems like a dangling sentence to be deleted (and not true since thin is an option)

mitzimorris replied:
good catch - removed.

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -58,39 +59,63 @@ All of these arguments are optional; when unspecified, the CmdStan defaults will
+**Get sample metadata**
+
+- ``column_names`` - List of column labels for one draw from the sampler.
+- ``sampler_vars_cols`` - Python dict, key: sampler parameter name, value: zero-based output column index.
```
`zero-based output column index` -> `tuple of output column indices`

mitzimorris replied:
fixed

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -58,39 +59,63 @@ All of these arguments are optional; when unspecified, the CmdStan defaults will
+- ``thin`` - Period between recorded iterations.
+
+- ``num_draws`` - Number of draws per chain, i.e., thinned iterations.
+- ``num_draws_sampling`` - Number of sampling (post-warmup) draws per chain, i.e., sampling iterations, thinned.  By default, only the post-warmup draws are reported, so that ``num_draws`` == ``num_draws_sampling``.
```
The `num_draws` behavior is a bug or this is wrong.

mitzimorris replied:
fixed

bbbales2 commented on docsrc/sample.rst:
```diff
@@ -58,39 +59,63 @@ All of these arguments are optional; when unspecified, the CmdStan defaults will
+
+- ``metric_type`` - Metric type used for adaptation, either ``diag_e``
+  or ``dense_e``, or ``None``, if the Stan program doesn't have any parameters.
+- ``num_params`` - Total number of parameters in the model; the sum of all scalar parameter variables and all elements of all container parameter variables
```
`parameters` -> `underlying unconstrained parameters`?

mitzimorris replied:
fixed

bbbales2 reviewed (changes requested):
I didn't pedantically check the tests, but looks like stuff is there. A few more comments/questions on docs/functions.

mitzimorris commented on cmdstanpy/stanfit.py:
```diff
@@ -364,57 +408,103 @@ def chain_ids(self) -> List[int]:
-        return self._draws_warmup + self._draws_sampling
+        """Number of draws per chain, i.e., thinned iterations."""
+        return int(
+            math.ceil((self._iter_sampling + self._iter_warmup) / self._thin)
```
right, I see how this is confusing - rename `num_draws` property to `num_draws_total`?  get rid of it?

my first concept for this interface was to keep warmup draws and sampling draws each in their own draws array.  but there's a use case for having a single array (don't remember what - for plots, maybe?).

the result is that default behavoir for `draws` is to return just the sample, plus optional arguent `inc_warmup`.

bbbales2 replied:
Hmm, I think just delete `fit.num_draws` is the simplest thing.

If someone wants the exact shape of the output of draws, they can use `.shape` of the output of draws. If they want the number of warmup draws or sampling draws for something else, they can use these functions.

mitzimorris replied:
deleted - execllent call

mitzimorris commented on cmdstanpy/stanfit.py:
```diff
@@ -364,57 +408,103 @@ def chain_ids(self) -> List[int]:
+    @property
+    def num_params(self) -> int:
+        """
+        Model parameters count.  This corresponds to the size of the
```
excellent point.  `num_unconstrained_params` - wordy, but accurate?

bbbales2 replied:
Wordy, but at least is a red flag to let the reader know there's something else going on if they care lol.

mitzimorris replied:
done!

mitzimorris commented:
hi @bbbales2 - made all changes requested.  renamed `num_params` to `num_unconstrained_params` and rewrote the doc to explain what's going on.

further cleanup:

- changed function `draws_pd` to use metadata from this refactor - I think this function is not useful and should be removed in the future.
- changed name of function `sampler_diagnostics` to `sampler_variables` to parallel function `stan_variables`.

bbbales2 reviewed (approved):
I fixed a couple typo things. Check em' and merge when you're happy!

mitzimorris commented:
awesome, many thanks!
